simulation/halsim_ws/src/main/native/cpp/HALSimWS.cpp

```diff
@@ -0,0 +1,123 @@
+#include "HALSimWS.h"
+#include "HALSimWebsocketConnection.h"
+#include "MessageSchema.h"
+
+#include <wpi/raw_ostream.h>
+#include <wpi/uv/util.h>
+
+namespace uv = wpi::uv;
+
+std::shared_ptr<HALSimWS> HALSimWS::g_instance;
+
+bool HALSimWS::Initialize() {
+  // Create the libuv loop
+  m_loop = uv::Loop::Create();
+  if (!m_loop) {
+    return false;
+  }
+
+  m_loop->error.connect([](uv::Error err) {
+    wpi::errs() << "uv ERROR: " << err.str() << "\n";
+  });
+
+  m_tcp_client = uv::Tcp::Create(m_loop);
+  if (!m_tcp_client) {
+    return false;
+  }
+
+  wpi::errs() << "HALSimWS Initialized\n";
+  return true;
+}
+
+void HALSimWS::Main(void* param) {
+  GetInstance()->MainLoop();
+  SetInstance(nullptr);
+}
+
+void HALSimWS::MainLoop() {
+  wpi::errs() << "Running MainLoop\n";
+  struct sockaddr_in dest;
+  uv::NameToAddr("localhost", 8080, &dest);
+
+  m_tcp_client->Connect(dest, [this, socket = m_tcp_client.get()]() {
```

Comment:
What happens if the connection fails? Seems like there should be some error handling for that?

Is there a reconnect capability? That seems important for a client implementation.

Related (and my implementation has the same issue), should probably figure out what the 'right' thing to do when the websocket disconnects. I lean towards setting the robot to disabled mode, but haven't really thought through it yet.

Reply:
Yeah I was planning on adding in reconnect capability with backoff (if we fail to connect, wait x number of seconds to retry, and if that fails wait x*2 seconds etc) for this.

Yeah we could signal the robot code that we are now in disabled state. This will probably let us play nicely if any other extensions (e.g. the GUI) are also loaded, since that should still be usable even if there's no websocket connection